diff --git a/.gitignore b/.gitignore
index ad1b63e2e9d6..cc6b4320d0f9 100644
--- a/.gitignore
+++ b/.gitignore
@@ -333,3 +333,6 @@ vs-chromium-project.txt
 
 # Ignore the default results output directory for tools/run-swarmed.py
 /results
+
+# Ignore qlabs
+qlabs/
\ No newline at end of file
diff --git a/BUILD.gn b/BUILD.gn
index aa2cb7b4583a..b40637f4312e 100644
--- a/BUILD.gn
+++ b/BUILD.gn
@@ -1490,3 +1490,12 @@ _lines = [
 assert(current_toolchain == default_toolchain)
 
 write_file("$root_build_dir/gn_logs.txt", _lines)
+
+group("qlabs_all") {
+  deps = [ "//qlabs" ]
+}
+
+group("qlabs_test_all") {
+  testonly = true
+  deps = [ "//qlabs:tests" ]
+}
diff --git a/net/server/BUILD.gn b/net/server/BUILD.gn
index 8ee35c5270f8..67ffa3f3f8a3 100644
--- a/net/server/BUILD.gn
+++ b/net/server/BUILD.gn
@@ -16,6 +16,7 @@ static_library("http_server") {
     "//chrome/browser/devtools",
     "//chrome/test/chromedriver/*",
     "//content/browser",
+    "//qlabs/*",
   ]
 
   friend = [
diff --git a/weblayer/BUILD.gn b/weblayer/BUILD.gn
index 3feb71cc9e39..58895aa0f5ec 100644
--- a/weblayer/BUILD.gn
+++ b/weblayer/BUILD.gn
@@ -372,6 +372,10 @@ source_set("weblayer_lib_base") {
     "renderer/weblayer_render_thread_observer.h",
     "utility/content_utility_client_impl.cc",
     "utility/content_utility_client_impl.h",
+
+    # echo
+    "//qlabs/learn_weblayer/echo/echo.cc",
+    "//qlabs/learn_weblayer/echo/echo.h",
   ]
 
   configs += [
@@ -540,6 +544,9 @@ source_set("weblayer_lib_base") {
     "//v8",
     "//weblayer:resources",
     "//weblayer/browser/webui:mojo_bindings",
+
+    # echo
+    "//qlabs/learn_weblayer/echo:mojo_bindings",
   ]
 
   if (enable_captive_portal_detection) {
@@ -900,7 +907,12 @@ grit("resources") {
     "grit/weblayer_resources.h",
     "weblayer_resources.pak",
   ]
-  deps = [ "//weblayer/browser/webui:mojo_bindings_webui_js" ]
+  deps = [
+    "//weblayer/browser/webui:mojo_bindings_webui_js",
+
+    # echo
+    "//qlabs/learn_weblayer/echo:mojo_bindings_js",
+  ]
 }
 # TODO(jam): move weblayer_shell_resources_grit and copy_shell_resources here in
 # a way that's shareable?
diff --git a/weblayer/browser/weblayer_browser_interface_binders.cc b/weblayer/browser/weblayer_browser_interface_binders.cc
index 6734c8d0c777..c1fcb3eda384 100644
--- a/weblayer/browser/weblayer_browser_interface_binders.cc
+++ b/weblayer/browser/weblayer_browser_interface_binders.cc
@@ -24,6 +24,9 @@
 #include "weblayer/browser/webui/weblayer_internals.mojom.h"
 #include "weblayer/browser/webui/weblayer_internals_ui.h"
 
+#include "qlabs/learn_weblayer/echo/echo.h"
+#include "qlabs/learn_weblayer/echo/echo.mojom.h"
+
 #if defined(OS_ANDROID)
 #include "mojo/public/cpp/bindings/self_owned_receiver.h"
 #include "services/service_manager/public/cpp/interface_provider.h"
@@ -73,6 +76,20 @@ void BindPageHandler(
   concrete_controller->BindInterface(std::move(receiver));
 }
 
+void BindEcho(content::RenderFrameHost* host,
+              mojo::PendingReceiver<echo::mojom::Echo> receiver) {
+  static qlabs::EchoImpl echo_impl;
+
+  auto* contents = content::WebContents::FromRenderFrameHost(host);
+  if (!contents)
+    return;
+
+  if (host->GetParent())
+    return;
+
+  echo_impl.BindInterface(std::move(receiver));
+}
+
 void BindNoStatePrefetchProcessor(
     content::RenderFrameHost* frame_host,
     mojo::PendingReceiver<blink::mojom::NoStatePrefetchProcessor> receiver) {
@@ -120,6 +137,8 @@ void PopulateWebLayerFrameBinders(
   map->Add<weblayer_internals::mojom::PageHandler>(
       base::BindRepeating(&BindPageHandler));
 
+  map->Add<echo::mojom::Echo>(base::BindRepeating(&BindEcho));
+
   map->Add<translate::mojom::ContentTranslateDriver>(
       base::BindRepeating(&BindContentTranslateDriver));
 
